# Remove VFS registration. Go handles it.
--- sqlite3.c.orig
+++ sqlite3.c
@@ -26884,7 +26884,7 @@
   sqlite3_free(p);
   return sqlite3_os_init();
 }
-
+#if 0 // Go handles VFS registration.
 /*
 ** The list of all registered VFS implementations.
 */
@@ -26981,7 +26981,7 @@
   sqlite3_mutex_leave(mutex);
   return SQLITE_OK;
 }
-
+#endif
 /************** End of os.c **************************************************/
 /************** Begin file fault.c *******************************************/
 /*
